firmware/main/main.ino

```diff
@@ -231,107 +231,104 @@ void Shutdown(bool with_flashing_leds = true) {
-            Serial.println("Wrong characteristic capabilities set on server");
-            client->disconnect();
-            break;
-          }
-          remote_characteristic->registerForNotify(ClientNotifyCallback);
-          connected_to_server = true;
-          Serial.printf("We are connected! Current seed: %u\n", remote_characteristic->readUInt32());
-        }
+bool EnsureConnectionToServer() {
+  if (last_device_connected) return true;
+
+  // The code below looks like it would leak memory like crazy...
+  BLEScanResults* found_devices = scan->start(BLE_SCAN_TIME_S, false);
+  Serial.print("Devices found: ");
+  Serial.println(found_devices->getCount());
+  for (int dev_id = 0; dev_id < found_devices->getCount(); ++dev_id) {
+    auto remote_device = found_devices->getDevice(dev_id);
+    BLEUUID service_uuid = remote_device.getServiceUUID();
+    if (service_uuid.equals(BLEUUID(SERVICE_UUID))) {
+      Serial.println("Found a device! Validating...");
+      client->setClientCallbacks(new MyClientCallback());
+      client->connect(&remote_device);
+      BLERemoteService* remote_service = client->getService(SERVICE_UUID); //do we ever need to free this?
+      if (remote_service == nullptr) {
+        Serial.println("We don't have the right service?");
+        client->disconnect();
+        continue;
       }
-      scan->clearResults();
-      delay(2000);
-    }
-    if (!connected_to_server) {
-      return;
-    } else {
-      // Wait for a sync point. We are doing a fast busy loop here to minimize latency.
-      while (true) {
-        if (have_new_value) {
-          uint32_t new_value_freshness = millis() - new_value_time;
-          // If we have a fresh value, we are done, otherwise discard and wait for the next sync point.
-          if (new_value_freshness < 20) {
-            break;
-          } else {
-            have_new_value = false;
-          }
-        }
-        delayMicroseconds(100);
+      BLERemoteCharacteristic* remote_characteristic = remote_service->getCharacteristic(CHARACTERISTIC_UUID);  //do we ever need to free this?
+      if (remote_characteristic == nullptr) {
+        Serial.println("We don't have the right characteristic?");
+        client->disconnect();
+        continue;
+      }
+      if (!remote_characteristic->canRead() || !remote_characteristic->canNotify()) {
+        Serial.println("Wrong characteristic capabilities set on server");
+        client->disconnect();
+        continue;
       }
-      have_new_value = false;
-      if (rng_state == 0) {
-        Serial.println("Remote glove says we are done. Shutting down.");
-        Shutdown();
+      remote_characteristic->registerForNotify(ClientNotifyCallback);
+      connected_to_server = true;
+      Serial.printf("We are connected! Current seed: %u\n", remote_characteristic->readUInt32());
+      break;
+    }
+  }
+  scan->clearResults();
+  delay(2000);
```

Comment:
We don't need to wait if we are now connected. This is just to insert some dead time between scans to save some battery. I guess this is part of why the connection is on startup is so slow (this accounts for 2 seconds!)